### PR TITLE
Upgrade to gradle 8.0.2, fix proguard rules for r8

### PR DESCRIPTION
**Pull request by VicV**
**Main problem**

The newest R8 strips a lot of stuff when its enabled.

I copied a ton of the details from here to add to our proguard consumer file:

https://github.com/square/retrofit/issues/3751

I also added a ton of protections around the jackson lib so that those types and files don't get stripped as well. 

Main item here too is making sure we `keep` the Android folder as it contains the coroutine schedule, objectmapper, and how we read configs locally and from the API

Other notes

* Force test apps to always run with R8 enabled so we catch these things earlier
* Update test apps to most recent version of gradle
* Update SDK to android 33 and newest Gradle 
* Update tests to handle new Android things which were failing

### CODE REVIEW

File: android-client-sdk/build.gradle [hunk 1]
```diff
@@ -24,7 +22,12 @@ android {
 
         //Will create field in buildconfig.java for use in code. 
         buildConfigField 'String', 'VERSION_NAME', "\"$versionName\""
+    }
+
+        namespace 'com.devcycle'
 
+    buildFeatures {
+        buildConfig = true
```
Comment (VicV) [inline android-client-sdk/build.gradle:30]: So this lets us have custom buildconfig props at least

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt [hunk 2]
```diff
@@ -48,7 +48,7 @@ internal data class PopulatedUser constructor(
     @Schema(description = "DevCycle SDK type")
     val sdkType: String = "mobile",
     @Schema(description = "DevCycle SDK Version")
-    val sdkVersion: String = BuildConfig.VERSION_NAME,
```
Comment (VicV) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt:51]: This is the problem place. 

This is forever null now. Idk how to pull the name from the config.

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt [hunk 3]
```diff
@@ -101,6 +101,7 @@ internal data class PopulatedUser constructor(
             val packageInfo: PackageInfo = packageManager.getPackageInfo(context.packageName, 0)
         
             val appVersion = packageInfo.versionName
+            val versionName = packageInfo.versionName;
```
Comment (VicV) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt:104]: this doesnt actually do anything and should be removed my bad

Comment (VicV) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/model/PopulatedUser.kt:104]: bump @laurawarr while youre in here

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/util/DVCSharedPrefs.kt [hunk 4]
```diff
@@ -2,7 +2,7 @@ package com.devcycle.sdk.android.util
 
 import android.content.Context
 import android.content.SharedPreferences
-import com.devcycle.sdk.android.R
+import com.devcycle.R
```
Comment (VicV) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/util/DVCSharedPrefs.kt:5]: I am worried this change will change the location of the prefs and thus the upgrade path is broken for folks without internet, need to test the upgrade path somehow with the same user id

File: java-example/gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -3,3 +3,6 @@ distributionPath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-7.4-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+
+## ENABLE THIS TO MAX OUT R8 MINIFICATION
+# android.enableR8.fullMode = true
```
Comment (VicV) [inline java-example/gradle/wrapper/gradle-wrapper.properties:8]: enable this too

File: kotlin-example/gradle/wrapper/gradle-wrapper.properties [hunk 6]
```diff
@@ -3,3 +3,7 @@ distributionPath=wrapper/dists
 distributionUrl=https\://services.gradle.org/distributions/gradle-7.4-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+
+## ENABLE THIS TO MAX OUT R8 MINIFICATION
```
Comment (VicV) [inline kotlin-example/gradle/wrapper/gradle-wrapper.properties:7]: enable this by default I guess

File: android-client-sdk/src/test/java/com/devcycle/sdk/android/api/DVCClientTests.kt [hunk 7]
```diff
@@ -383,38 +383,6 @@ class DVCClientTests {
...
-    fun `variable calls back when variable value changes`() {
-        val config = generateConfig("activate-flag", "Flag activated!", Variable.TypeEnum.STRING)
-
-        mockWebServer.enqueue(MockResponse().setResponseCode(200).setBody(objectMapper.writeValueAsString(config)))
-
-        val client = createClient("pretend-its-a-real-sdk-key", mockWebServer.url("/").toString())
-
-        try {
-            val variable = client.variable("activate-flag", "Not activated")
-            variable.onUpdate(object: DVCCallback<Variable<String>> {
```
Comment (laurawarr) [inline android-client-sdk/src/test/java/com/devcycle/sdk/android/api/DVCClientTests.kt:396]: @VicV this approach for onUpdate seems to be deprecated in favour of using a plain function (see test below), removing this has all tests passing. I believe this is related to the bump to java 18

Comment (VicV) [reply android-client-sdk/src/test/java/com/devcycle/sdk/android/api/DVCClientTests.kt:396]: @laurawarr I don't have any context around this test / onupdate function im just a man that plays around with gradle you might want a diff reviewer on that

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/util/DVCSharedPrefs.kt (hunk 4 above)
Comment (laurawarr) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/util/DVCSharedPrefs.kt:5]: Tested the upgrade path by: 
- loading test app with latest released version of the SDK
- reloading app with local version of SDK
- verified that user config is initially loaded from cache

**Review by kaushalkapasi — APPROVED** [review]

**Pull request merged**